### ISSUE TITLE
multi class text classifier 

### ISSUE THREAD

**Issue by azarezade** [explosion/spaCy#1997]
I have problem deciding which way is better to use for multi-class text-classification. For example, the output of the following two class problem:
```Python
import spacy
nlp = spacy.load('en')

train_data = [
    (u"That was very bad", {"cats": {"POSITIVE": 0, "NEGATIVE": 1}}),
    (u"it is so bad", {"cats": {"POSITIVE": 0, "NEGATIVE": 1}}),
    (u"so terrible", {"cats": {"POSITIVE": 0, "NEGATIVE": 1}}),
    (u"I like it", {"cats": {"POSITIVE": 1, "NEGATIVE": 0}}),
    (u"It is very good.", {"cats": {"POSITIVE": 1, "NEGATIVE": 0}}),
    (u"That was great!", {"cats": {"POSITIVE": 1, "NEGATIVE": 0}})
]

textcat = nlp.create_pipe('textcat')
nlp.add_pipe(textcat, last=True)
textcat.add_label('POSITIVE')
textcat.add_label('NEGATIVE')
optimizer = nlp.begin_training()
for itn in range(10):
    for doc, gold in train_data:
        nlp.update([doc], [gold], sgd=optimizer)
doc = nlp(u'It is good.')
print(doc.cats)
doc = nlp(u'It is bad.')
print(doc.cats)
```
is
```
{'POSITIVE': 0.9940504431724548, 'NEGATIVE': 0.03267182409763336}
{'POSITIVE': 0.6032131314277649, 'NEGATIVE': 0.047026827931404114}
```
but when I use 
```Python
train_data = [
    (u"That was very bad", {"cats": {"NEGATIVE": 1}}),
    (u"it is so bad", {"cats": {"NEGATIVE": 1}}),
    (u"so terrible", {"cats": {"NEGATIVE": 1}}),
    (u"I like it", {"cats": {"POSITIVE": 1}}),
    (u"It is very good.", {"cats": {"POSITIVE": 1}}),
    (u"That was great!", {"cats": {"POSITIVE": 1}})
]
```
the result is different:
```
{'POSITIVE': 0.9999344348907471, 'NEGATIVE': 0.06132430210709572}
{'POSITIVE': 0.9991464614868164, 'NEGATIVE': 0.21219134330749512}
```
I don't think this is due to randomization, since in different runs of the same code the results remains more of less the same.

I want to know, which one is the correct way to define the training data in multi-class `textcat`?

## Environment
* Operating System: Mac OS X
* Python Version Used: 3.6
* spaCy Version Used: 2.0.7

**Comment by honnibal** [comment]
If your classes are mutually exclusive, the first way is correct. If the model doesn't have a label for an instance/class pair, it regards the value as unknown.

I need to add an option to the text classifier to do mutually exclusive prediction.

**Comment by azarezade** [comment]
Thanks Matt!
That would be very good option to add, since in most applications there are many mutually exclusive classes, indeed.

**Comment by skrcode** [comment]
@honnibal Instead of keeping the option for multiclass, what about adding a value of 0 to all the labels not present. If any of the labels are not specified, the intent is to have the missing ones as 0. Won't this solve for both the multiclass and the multilabel problem? Please correct me if this assumption is incorrect. 
https://github.com/explosion/spaCy/blob/master/spacy/language.py
     
        # Allow dict of args to GoldParse, instead of GoldParse objects.
        gold_objs = []
        doc_objs = []
        
        # Add not found labels to gold
        for gold in golds:          
            for name, proc in list(self.pipeline):
                if isinstance(proc, TextCategorizer):
                    for j, label in enumerate(proc.labels):
                        if label not in gold['cats']:
                            gold['cats'][label]=0

        for doc, gold in zip(docs, golds):
            if isinstance(doc, basestring_):
                doc = self.make_doc(doc)
            if not isinstance(gold, GoldParse):
                gold = GoldParse(doc, **gold)
            doc_objs.append(doc)
            gold_objs.append(gold)

**Comment by mohitkrpopxo** [comment]
Hey @azarezade I am working on a similar problem but with many classes, 35 to be specific. My training sample is 10000 rows overall. And I have fed the data in the first way, i.e.
train_data = [
    (u"That was very bad", {"cats": {"POSITIVE": 0, "NEGATIVE": 1}}),
    (u"it is so bad", {"cats": {"POSITIVE": 0, "NEGATIVE": 1}}),
    (u"so terrible", {"cats": {"POSITIVE": 0, "NEGATIVE": 1}}),
    (u"I like it", {"cats": {"POSITIVE": 1, "NEGATIVE": 0}}),
    (u"It is very good.", {"cats": {"POSITIVE": 1, "NEGATIVE": 0}}),
    (u"That was great!", {"cats": {"POSITIVE": 1, "NEGATIVE": 0}})
]

However, I am not getting accurate prediction in this case.

When I was working with four mutually exclusive classes, the model was giving very good results with fewer training sample. Any idea why this might happen?

**Comment by azarezade** [comment]
Maybe your model couldn't train well because of your dataset. Actually there could be many reason for that, like unbalanced training data and so on.

Since you could learn the model successfully with 4 classes, it shows that you training setup (I mean the way to represent your data) is ok.

**Comment by mohitkrpopxo** [comment]
When I trained the model with 4 classes, the classes were exclusive for a document. However, in the case of ~40 classes, each document had multiple classes. My Sample training input looks like:
{'Astro World': False,
 'Bath & Body': False,
 'Beauty': False,
 'Celebrations': False,
 'Celebrity Weddings': False,
 'DIY': False,
 'Entertainment': False,
 'Essentials': False,
 'Face Care': False,
 'Face Care Products': False,
 'Fashion Accessories': False,
 'Fashion Tips & Trends': False,
 'Food': False,
 'Hair Care': False,
 'Health': False,
 'Health & Fitness': False,
 'Home & Garden': False,
 'Indian': False,
 'Indian Wear Trends': False,
 'Make Up': False,
 'Makeup Tips': False,
 'Nails': False,
 'Products': False,
 'Relationships': False,
 'Sex': False,
 'Skin Care': False,
 'Skin Care Products': False,
 'Skin Care Tips': False,
 'Travel': False,
 'Wedding Beauty': False,
 'Wedding Fashion': True,
 'Wedding Ideas & Tips': False,
 'Wedding Lifestyle': True,
 'Western': False,
 'Western Wear Trends': False,
 'Work': False}

The above text had 2 categories: 'Wedding Ideas & Tips' and 'Wedding Fashion'.

Is this the right way to input training data?
`
def load_data(limit=0, split=0.8):
    """Load data from dataset."""
    # Partition off part of the train data for evaluation
    all_data = pd.read_csv('~/Downloads/subcategories.csv')
    test_data = all_data[2000:]
    train_data = all_data[0:2000]
    train_data = train_data.values.tolist()
    random.shuffle(train_data)
    train_data = train_data[-limit:]
    texts, labels = zip(*train_data)
    
    # Conditions for multi label input
    default_dic = {'Astro World': False, 'Bath & Body': False, 'Beauty': False, 'Celebrations': False, 'Celebrity Weddings': False, 'DIY': False, 'Essentials': False, 'Entertainment': False, 'Face Care': False, 'Face Care Products': False, 'Fashion Accessories': False, 'Fashion Tips & Trends': False, 'Food': False, 'Hair Care': False, 'Health': False, 'Health & Fitness': False, 'Home & Garden': False, 'Indian': False, 'Indian Wear Trends': False, 'Make Up': False, 'Makeup Tips': False, 'Nails': False, 'Products': False, 'Relationships': False, 'Sex': False, 'Skin Care': False, 'Skin Care Products': False, 'Skin Care Tips': False, 'Travel': False, 'Wedding Beauty': False, 'Wedding Fashion': False, 'Wedding Ideas & Tips': False, 'Wedding Lifestyle': False, 'Western': False, 'Western Wear Trends': False, 'Work': False}
    cats = []
    for l in labels:
        ls = l.split(',')
        for lab in ls:
            if lab in default_dic:
                default_dic[lab]=True
        new_dic = dict(default_dic)
        cats.append(new_dic)
        for lab in ls:
            if lab in default_dic:
                default_dic[lab]=False

    split = int(len(train_data) * split)
    return (texts[:split], cats[:split]), (texts[split:], cats[split:])
`

**Comment by svlandeg** [comment]
I think this can be closed as the original issue was resolved by [adding](https://github.com/svlandeg/spaCy/commit/e9dd5943b97b6eb9fee767d00cd85250440375d7) a specific option to set `exclusive_classes` to `True`, which then results in a `Softmax` output layer. From spaCy 3.0 onwards, it will be even easier to change these kind of settings in a configuration file.
